pkg/git/git.go

```diff
@@ -0,0 +1,53 @@
+package git
+
+import (
+	"github.com/fluxcd/source-controller/pkg/git/gogit"
+	extgogit "github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+
+	"github.com/go-git/go-git/v5/config"
+	"github.com/go-git/go-git/v5/storage/memory"
+	corev1 "k8s.io/api/core/v1"
+)
+
+//Client git client
+type Client struct{}
+
+//ListTags returns a list of tags for a given repository
+func (c *Client) ListTags(url string, secret *corev1.Secret) ([]string, error) {
+	rem := extgogit.NewRemote(memory.NewStorage(), &config.RemoteConfig{
+		Name: "origin",
+		URLs: []string{url},
+	})
+
+	var auth transport.AuthMethod
+	if secret != nil {
+		authStrategy, err := gogit.AuthSecretStrategyForURL(url)
+
+		if err != nil {
+			return nil, err
+		}
+		authMethod, err := authStrategy.Method(*secret)
+		if err != nil {
+			return nil, err
+		}
+		auth = authMethod.AuthMethod
+	}
+
+	refs, err := rem.List(&extgogit.ListOptions{
+		Auth: auth,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	var tags []string
+	for _, ref := range refs {
+		if ref.Name().IsTag() {
+			tags = append(tags, ref.Name().Short())
+		}
+	}
+
+	return tags, nil
+}
```

Comment:
this doesn't have any unit tests :sob: I considered adding some, but I'd end up mocking out like 5/6 single-function interfaces. This logic gets tested as part of the acceptance tests

Reply:
you _could_ have a single interface with N funcs which delegate to each real function 🤷‍♀️